spec/app/ics-030-middleware/README.md

```diff
@@ -0,0 +1,190 @@
+
+Without a middleware approach, developers must choose whether to place this extension to application logic inside each relevant application; or place the logic in core IBC. Placing it in each application is redundant and prone to error. Placing the logic in core IBC requires an opt-in from all applications and violates the abstraction barrier between core IBC (tao) and the application. Either case is not scalable as the number of extensions increase, since this must either increase code bloat in applications or core IBC handlers.
+
+Middleware allows developers to define the extensions as seperate modules that can wrap over the end application. This middleware can thus perform its own custom logic, and pass data into the application so that it may run its logic without being aware of the middleware's existence. This allows both the application and the middleware to implement its own isolated logic while still being able to run as part of a single packet flow.
+
+### Desired Properties
+
+- Middleware enables arbitrary extensions of application logic
+- Middleware can be arbitrarily nested to create a chain of app extensions
+- Core IBC does not need to change
+- Base Application logic does not need to change
+
+## Technical Specification
+
+### General Design
+
+In order to function as IBC Middleware, a module must implement the IBC application callbacks and pass along the pre-processed data to the nested application. It must also implement `WriteAcknowledgement` and `SendPacket`, which will be called by the end application, so that it may post-process the information before passing data along to core ibc.
+
+When nesting an application, the module must make sure that it is in the middle of communication between core IBC and the application in both directions. Developers should do this by registering the top-level module directly with the IBC router (not any nested applications). The nested applications in turn, must be given access only to the middleware's `WriteAcknowledgement` and `SendPacket` rather than to the core IBC handlers directly.
+
+Additionally, the middleware must take care to ensure that the application logic can execute its own port and version negotiation without interference from the nesting middleware. In order to do this, the middleware will prepend the portID and version with its own portID and version. In the application callbacks, the middleware must do its own version and port negotation and then strip out the prefixes before handing over the data to the nested application's callback. Middleware SHOULD always prepend the portID with its own port. This will allow the original application to also exist as a top-level module connected to the IBC Router.
+
+PortID: `{middleware_port}:{app_port}`
```

Comment:
On the other hand, this seems very odd to me. We can have any port for an app.
`ics20` could listen on `transfer` or `custom` or `wasm.cosmos1dj8fifj3ljfliw7nfh8w87cb` (I have tests with all these in `ts-relayer`).

The application logic has no hardcoded port and should not concern itself with that. That should be a param that can be set in genesis (or elsewhere). 

Basically, I would remain that port can be configured to anything and remove this line, just saying that different middleware stacks would be different ports.

You could even do `transfer` => `fee-1:ics20-1`
 and `old-transfer` => `ics20-1`

Forcing people to connect `old-transfer` with older chains.